basesandbox/src/execution/executor.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::ipc::*;
+use std::collections::HashMap;
+use std::process::Command;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+pub trait Executor {
+    fn new(path: &str, args: &[&str]) -> Self;
+    fn join(&mut self);
+}
+
+pub struct Executable {
+    child: std::process::Child,
+}
+
+impl Executor for Executable {
+    fn new(path: &str, args: &[&str]) -> Self {
+        let mut command = Command::new(path);
+        command.args(args);
+        Executable {
+            child: command.spawn().unwrap(),
+        }
+    }
+
+    fn join(&mut self) {
+        // This is synchronized with excutee's signal (#TERMINATE),
+        // which is supposed to be sent in termination step.
+        // Thus, in normal case, it won't take much time to be in a waitable status.
+        // However a malicous excutee might send a signal arbitrarily before the termination.
+        // For that, we have a timeout for the wait.
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(50));
+            if self.child.try_wait().unwrap().is_some() {
+                return // successful
+            }
+        }
+        panic!("Module hasn't terminated itself. Protocol Error")
+    }
+}
+
+/// This is for the SameProcess tasks. You can register a runnable function,
+/// and it will be executed later as an instance of 'process' (which is actually not).
+pub type ThreadAsProcesss = Arc<dyn Fn(Vec<String>) -> () + Send + Sync>;
+lazy_static! {
```

Comment:
I recommend you to use [`once_cell`](https://crates.io/crates/once_cell) instead of `lazy_static`. It's currently in the process of being accepted into the `std`. It doesn't use macros at all.

Reply:
Good to know. Thanks